### PR TITLE
cache data of exernal workbook 

### PR DESCRIPTION
**Pull request by aspojo**
cache data of exernal workbook ,so that excel can show the cached cell value when the linked workbook not exists

### CODE REVIEW

File: poi-ooxml/src/main/java/org/apache/poi/xssf/usermodel/BaseXSSFFormulaEvaluator.java [hunk 1]
```diff
@@ -26,10 +28,12 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 import org.apache.poi.ss.formula.eval.NumberEval;
 import org.apache.poi.ss.formula.eval.StringEval;
 import org.apache.poi.ss.formula.eval.ValueEval;
-import org.apache.poi.ss.usermodel.Cell;
-import org.apache.poi.ss.usermodel.CellType;
```
Comment (pjfanning) [inline poi-ooxml/src/main/java/org/apache/poi/xssf/usermodel/BaseXSSFFormulaEvaluator.java:30]: can you keep the full imports as opposed to using wildcards?

File: poi-ooxml/src/main/java/org/apache/poi/xssf/model/ExternalLinksTable.java [hunk 2]
```diff
@@ -33,10 +33,7 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 import org.apache.poi.util.Internal;
 import org.apache.poi.util.Removal;
 import org.apache.xmlbeans.XmlException;
-import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTExternalDefinedName;
-import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTExternalLink;
-import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTExternalSheetName;
```
Comment (pjfanning) [inline poi-ooxml/src/main/java/org/apache/poi/xssf/model/ExternalLinksTable.java:38]: can you keep the full imports as opposed to using wildcards?

File: poi-ooxml/src/main/java/org/apache/poi/xssf/usermodel/BaseXSSFFormulaEvaluator.java [hunk 3]
```diff
@@ -73,9 +78,60 @@ protected CellValue evaluateFormulaCellValue(Cell cell) {
...
+                                XSSFCell cell = row.getCell(cellIndex);
+                                String cellValue = cell.getRawValue();
+                                String cellR = new CellReference(cell).formatAsString(false);
+                                externalLinksTable.cacheData(sheet.getSheetName(), rowIndex + 1, cellR, cellValue);
+
+                            }
+                        }
+
+                    }
+                    System.out.println("externalLinksTable = " + externalLinksTable);
```
Comment (pjfanning) [inline poi-ooxml/src/main/java/org/apache/poi/xssf/usermodel/BaseXSSFFormulaEvaluator.java]: Please remove the system outs or replace with log4j debug statements

File: poi-ooxml/src/test/java/org/apache/poi/xssf/usermodel/TestXSSFWorkbook.java [hunk 4]
```diff
@@ -60,11 +61,7 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 import org.apache.poi.xssf.XSSFITestDataProvider;
 import org.apache.poi.xssf.model.StylesTable;
 import org.junit.jupiter.api.Test;
-import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTCalcPr;
-import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTPivotCache;
-import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTWorkbook;
-import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTWorkbookPr;
```
Comment (pjfanning) [inline poi-ooxml/src/test/java/org/apache/poi/xssf/usermodel/TestXSSFWorkbook.java:66]: again, keep the non-wildcard imports

File: poi-ooxml/src/test/java/org/apache/poi/xssf/usermodel/TestXSSFWorkbook.java [hunk 5]
```diff
@@ -1299,6 +1296,50 @@ void testIgnoreTrashParts() throws Exception {
...
+            String formula = String.format(LocaleUtil.getUserLocale(), "SUM('[%s]Sheet0'!A1:B1)", nameA);
+            cell.setCellFormula(formula);
+            XSSFFormulaEvaluator evaluator = workbookB.getCreationHelper().createFormulaEvaluator();
+            evaluator.evaluateFormulaCell(cell);
+
+            assertEquals(sum, cell.getNumericCellValue());
+
+            workbookA.write(bosA);
+            workbookB.write(bosB);
+            FileOutputStream f1 = new FileOutputStream(nameA);
```
Comment (pjfanning) [inline poi-ooxml/src/test/java/org/apache/poi/xssf/usermodel/TestXSSFWorkbook.java]: please avoid writing files in tests, use UnsynchronizedByteArrayOutputStream instead

**Review by pjfanning — CHANGES_REQUESTED** [review]

**Review by aspojo — COMMENTED** [review]
all resolved.

**Comment by pjfanning** [comment]
thanks - merged with https://github.com/apache/poi/commit/68ce022531f47497226f42f434cc8a78c000ebbe

**Pull request closed without merging**